wandb/sdk/launch/runner/kubernetes_monitor.py

```diff
@@ -0,0 +1,297 @@
+
+        self._watch_job_thread = Thread(target=self._watch_job, daemon=True)
+        self._watch_pods_thread = Thread(target=self._watch_pods, daemon=True)
+        self._watch_crd_thread = Thread(target=self._watch_crd, daemon=True)
+
+        self._job_watcher = SafeWatch(watch.Watch())
+        self._pod_watcher = SafeWatch(watch.Watch())
+
+    def start(self) -> None:
+        """Start the run monitor."""
+        if self.custom_api is None:
+            self._watch_job_thread.start()
+        else:
+            self._watch_crd_thread.start()
+        self._watch_pods_thread.start()
+
+    def stop(self) -> None:
+        """Stop the run monitor."""
+        self._job_watcher.stop()
+        self._pod_watcher.stop()
+
+    def _set_status(self, status: Status) -> None:
+        """Set the run status."""
+        with self._status_lock:
+            self._status = status
+
+    def get_status(self) -> Status:
+        """Get the run status."""
+        with self._status_lock:
+            # Each time this is called we verify that our watchers are active.
+            if self._status.state in ["running", "starting"]:
+                if self.custom_api is None:
+                    if not self._watch_job_thread.is_alive():
+                        wandb.termwarn(
+                            f"Job watcher thread is dead for {self.job_field_selector}"
+                        )
+                        self._watch_job_thread = Thread(
+                            target=self._watch_job, daemon=True
+                        )
+                        self._watch_job_thread.start()
+                else:
+                    if not self._watch_crd_thread.is_alive():
+                        wandb.termwarn(
+                            f"CRD watcher thread is dead for {self.job_field_selector}"
+                        )
+                        self._watch_crd_thread = Thread(
+                            target=self._watch_crd, daemon=True
+                        )
+                        self._watch_crd_thread.start()
+                if not self._watch_pods_thread.is_alive():
+                    wandb.termwarn(
+                        f"Pod watcher thread is dead for {self.pod_label_selector}"
+                    )
+                    self._watch_pods_thread = Thread(
+                        target=self._watch_pods, daemon=True
+                    )
+                    self._watch_pods_thread.start()
+            return self._status
+
+    def _watch_pods(self) -> None:
+        """Watch for pods created matching the jobname."""
+        # Stream with no timeout polling for pod status updates
+        for event in self._pod_watcher.stream(
+            self.core_api.list_namespaced_pod,
+            namespace=self.namespace,
+            label_selector=self.pod_label_selector,
+        ):
+            type = event.get("type")
+            object = event.get("object")
+
+            if type == "MODIFIED":
```

Comment:
I think we need to get rid of this check and treat `ADDED` events exactly as if they were `MODIFIED` events